crates/sdk/src/masp.rs

```diff
@@ -622,35 +618,123 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
+                 decrypt notes with"
+                    .to_string(),
+            ));
+        };
+        let maybe_last_witnessed_tx = self.tx_note_map.keys().max().cloned();
+        let last_height_in_witnesses = std::cmp::min(
+            maybe_last_witnessed_tx.as_ref(),
+            maybe_least_synced_vk_height.as_ref(),
+        )
+        .map(|ix| ix.height);
+        Ok(last_height_in_witnesses.unwrap_or_else(BlockHeight::first))
+    }
+
+    #[cfg(not(target_family = "wasm"))]
+    async fn update_with_pre_built_data<M>(
+        &mut self,
+        client: &M,
+        height: BlockHeight,
+    ) -> Result<(), Error>
+    where
+        M: MaspClient,
+    {
+        trait ThenAwait {
+            #[allow(async_fn_in_trait)]
+            async fn then_await<F>(self, fut: F) -> Option<F::Output>
+            where
+                F: std::future::Future;
+        }
+
+        impl ThenAwait for bool {
+            async fn then_await<F>(self, fut: F) -> Option<F::Output>
+            where
+                F: std::future::Future,
+            {
+                if self { Some(fut.await) } else { None }
+            }
+        }
+
+        let tree_fut = async {
+            client
+                .capabilities()
+                .may_fetch_pre_built_tree()
+                .then_await(client.fetch_commitment_tree(height))
+                .await
+                .transpose()
+        };
+        let notes_map_fut = async {
+            client
+                .capabilities()
+                .may_fetch_pre_built_notes_map()
+                .then_await(client.fetch_tx_notes_map(height))
+                .await
+                .transpose()
+        };
+        let witness_map_fut = async {
+            client
+                .capabilities()
+                .may_fetch_pre_built_witness_map()
+                .then_await(client.fetch_witness_map(height))
+                .await
+                .transpose()
+        };
+
+        let (maybe_tree, maybe_notes_map, maybe_witness_map) =
+            futures::try_join!(tree_fut, notes_map_fut, witness_map_fut)?;
+
+        if let Some(tree) = maybe_tree {
+            self.tree = tree;
+        }
+        if let Some(notes_map) = maybe_notes_map {
+            self.tx_note_map = notes_map;
+        }
+        if let Some(witness_map) = maybe_witness_map {
+            self.witness_map = witness_map;
+        }
+
+        Ok(())
+    }
+
     #[allow(clippy::too_many_arguments)]
     #[cfg(not(target_family = "wasm"))]
-    async fn fetch_aux<'client, C, IO, M>(
+    async fn fetch_aux<IO, M>(
         &mut self,
         client: M,
         progress: &impl ProgressTracker<IO>,
         start_query_height: Option<BlockHeight>,
         last_query_height: Option<BlockHeight>,
         retry: RetryStrategy,
-        // NOTE: do not remove this argument, it will be used once the indexer
-        // is ready
-        _batch_size: u64,
```

Comment:
Is it ok to remove this arg? I believe @batconjurer had some plans for it but I'm not up to date with the latest decisions for the indexer

Reply:
I'm good with it because of the phase 3 design that we have now.